scripts/tests/run_tv_casting_test.py

```diff
@@ -0,0 +1,245 @@
+DEVICE_TYPE = 35
+
+
+# Dump the logs to the console in the case of an error.
+def dump_logs_to_console(log_file):
+
+    if log_file == LINUX_TV_CASTING_APP_LOGS:
+        logging.debug('Dumping Linux TV Casting App Logs to Console.')
+    elif log_file == LINUX_TV_APP_LOGS:
+        logging.debug('Dumping Linux TV App Logs to Console.')
+
+    with open(log_file, 'r') as file:
+        logs = file.read()
+        logging.debug(logs)
+
+
+# Remove the log files once the script is done running.
+def remove_log_file(log_file):
+
+    if os.path.exists(log_file):
+        os.remove(log_file)
+    else:
+        logging.error('The file does not exist.')
+
+
+# Whenever a failure is discovered, we should print 'Discovery failed!',
+# dump the logs, clean up the log files, exit on error.
+def handle_discovery_failure():
+
+    logging.error('Discovery failed!\n')
+
+    dump_logs_to_console(LINUX_TV_CASTING_APP_LOGS)
+    dump_logs_to_console(LINUX_TV_APP_LOGS)
+
+    remove_log_file(LINUX_TV_CASTING_APP_LOGS)
+    remove_log_file(LINUX_TV_APP_LOGS)
+
+    sys.exit(1)
+
+
+# Helper function to extract the integer value from a string.
+def extract_value_from_string(line):
+
+    value = line.split(':')[-1].strip().replace('\x1b[0m', '')
+    value = int(value)
+
+    return value
+
+
+# Check if the discovered value matches the expected value.
+# Returns False if the value does not match, True otherwise.
+def validate_value(expected_value, line, value_name):
+
+    # Extract the integer value from the string.
+    value = extract_value_from_string(line)
+
+    # If the discovered value does not match the expected value,
+    # print the error and return False.
+    if value != expected_value:
+        logging.error(f'{value_name} does not match the expected value!')
+        logging.error(f'Expected {value_name}: {expected_value}')
+        line = line.rstrip('\n')
+        logging.error(line)
+
+        return False
+
+    # Return True if the value matches the expected value.
+    return True
+
+
+# Tear down the processes once done running.
+def tear_down(processes):
+
+    for process in processes:
+        process.terminate()
+        process.wait()
+
+
+# Test if the Linux tv-casting-app is able to discover the Linux tv-app.
+# The Linux tv-casting-app and the tv-app will be run in separate processes.
+# Their corresponding output will be written to their respective log files.
+# The output of the tv-casting-app will be parsed in realtime for strings of
+# interest which will be printed to the console.
+def test_discovery_fn():
+
+    with open(LINUX_TV_APP_LOGS, 'w') as fd1, open(LINUX_TV_CASTING_APP_LOGS, 'w') as fd2:
+
+        # Run the Linux tv-app and write the output to file.
+        tv_app_rel_path = 'out/tv-app/chip-tv-app'
+        tv_app_abs_path = os.path.abspath(tv_app_rel_path)
+        tv_app_process = subprocess.Popen(tv_app_abs_path, stdout=fd1, stderr=subprocess.PIPE, text=True)
+
+        time.sleep(RUN_INTERVAL)
```

Comment:
Could we instead have some "wait max X seconds until condition is met ..." instead?

We seem to sleep and then assume everything is well. Generally Sleep() in tests is not a good idea. We should poll the output in this case it seems and check if it contains some signal of "program started successfully" and time out if we do not see the expected output.

A "sleep the maximum amount of program may take to start" will always be pessimistic and will make the tests as slow as possible.

Reply:
Noted! I changed to this approach in the latest commit.